docs: publish example controllers and service as @metamask/sample-controllers

dbrans opened the pull request:
## Explanation

Publish existing sample controllers as @metamask/sample-controllers so it can be used in sample code.


## References
* Related to https://github.com/MetaMask/MetaMask-planning/issues/3850
<!--
Are there any issues that this pull request is tied to?
Are there other links that reviewers should consult to understand these changes better?
Are there client or consumer pull requests to adopt any breaking changes?

For example:

* Fixes #12345
* Related to #67890
-->

## Changelog

<!--
If you're making any consumer-facing changes, list those changes here as if you were updating a changelog, using the template below as a guide.

(CATEGORY is one of BREAKING, ADDED, CHANGED, DEPRECATED, REMOVED, or FIXED. For security-related issues, follow the Security Advisory process.)

Please take care to name the exact pieces of the API you've added or changed (e.g. types, interfaces, functions, or methods).

If there are any breaking changes, make sure to offer a solution for consumers to follow once they upgrade to the changes.

Finally, if you're only making changes to development scripts or tests, you may replace the template below with "None".
-->

### `@metamask/sample-controllers`

- **ADDED**: Initial sample controllers and service – (GasPricesController, PetNamesController, GasPricesService)

## Checklist

- [ ] I've updated the test suite for new or updated code as appropriate
- [ ] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [ ] I've highlighted breaking changes using the "BREAKING" category above as appropriate
- [ ] I've prepared draft pull requests for clients and consumer packages to resolve any breaking changes

dbrans commented:
@metamaskbot publish-preview

dbrans commented:
@metamaskbot publish-preview

dbrans commented:
@metamaskbot publish-preview

mcmire commented:
Hi @dbrans can you provide more information on why these need to be published? These really are just example controllers -- they aren't hooked up to any features in any of the clients. They were originally created as templates people could follow when making new controllers. How do you envision these packages being used?

dbrans commented:
Hi @mcmire I pinged you on slack to discuss offline.

mcmire commented on packages/sample-controllers/package.json:
Since GasPricesController talks to NetworkController, can we add `@metamask/network-controller` to both `devDependencies` and `peerDependencies`?

dbrans replied:
Done.

mcmire commented on packages/sample-controllers/tsconfig.json:
Since GasPricesController uses a type from NetworkController, can we add `../../packages/network-controller` to the `references` field in this file?

dbrans replied:
Done.

mcmire commented on packages/sample-controllers/tsconfig.build.json:
Since GasPricesController uses a type from NetworkController, can we add `../../packages/network-controller/tsconfig.build.json` to the `references` field in this file?

dbrans replied:
Done.

mcmire reviewed (commented):
Had some suggestions, let me know what you think. Can you run `yarn update-readme-content` after you've made the changes to `package.json`?

dbrans commented:
@mcmire: I added the "Sample" prefix and renamed to avoid confustion in https://github.com/MetaMask/core/pull/5363/commits/d96e665b9bdbbbc4dedf48ee82a91b28d09a183b

mcmire commented on packages/sample-controllers/src/sample-gas-prices-service/abstract-gas-prices-service.ts:
What are your thoughts on renaming this one to `SampleAbstractGasPricesService`?

dbrans replied:
Done.

mcmire commented on packages/sample-controllers/src/sample-gas-prices-service/index.ts:
```diff
@@ -0,0 +1,2 @@
+export type * from './abstract-gas-prices-service';
```
Hmm, it'd be nice if we could avoid barrel exports. That's something I'm trying to discourage people from using because it makes it difficult to understand what exactly is being exported. It's technically okay at this layer — it's really `src/index.ts` that shouldn't have the barrel exports — but I feel like once people see it in use they will think they can use it everywhere. Thoughts on removing this file since these are supposed to be examples?

dbrans replied:
Done.

dbrans replied:
Done.

mcmire reviewed (commented):
A couple of things but this otherwise looks good.

mcmire commented on packages/sample-controllers/src/index.ts:
It looks like the coverage tool is saying that this file is untested. For other repos we've fixed this by adding an `index.test.ts` that at least verifies the JavaScript exports. You can see an example of this here: https://github.com/MetaMask/core/blob/931ba41b9c3af3b87a51f9b24d6c2221e084094a/packages/controller-utils/src/index.test.ts. Perhaps we could add that?

dbrans replied:
We ignore 'src/index.ts' from coverage in our jest tests:

https://github.com/MetaMask/core/blob/main/jest.config.packages.js#L31

The index.ts that is failing coverage is the one in src/sample-gas-prices-service. 

I added tests for all the index.ts files in this package. However, it is still showing 0% Func coverage. There aren't any functions defined in that file.

I've modified the ignore path to ingore all index.ts files. Let me know what you think.

mcmire reviewed (commented):
Sorry one more thing!

mcmire reviewed (approved):
Looks good!
